File: src/main/java/net/easecation/ghosty/recording/player/updated/PlayerUpdatedWorldChanged.java
package net.easecation.ghosty.recording.player.updated;

import cn.nukkit.Server;
import cn.nukkit.level.Location;
import cn.nukkit.utils.BinaryStream;
import net.easecation.ghosty.entity.PlaybackNPC;
import net.easecation.ghosty.recording.player.PlayerRecordNode;

import java.util.Objects;

/**
 * Created by Mulan Lin('Snake1999') on 2016/11/19 15:23.
 */
public class PlayerUpdatedWorldChanged implements PlayerUpdated {

    @Override
    public int getUpdateTypeId() {
        return TYPE_WORLD;
    }

    public static PlayerUpdatedWorldChanged of(String tn) {
        return new PlayerUpdatedWorldChanged(tn);
    }

    @Override
    public boolean hasStates() {
        return true;
    }

    private String wn;

    @Override
    public void processTo(PlaybackNPC ghost) {
        Location location = ghost.getLocation();
        location.level = Server.getInstance().getLevelByName(wn);
        if (location.level == null) return;
        // ghost.teleport(location);
    }

    @Override
    public PlayerRecordNode applyTo(PlayerRecordNode node) {
        node.setLevel(wn);
        return node;
    }

    public PlayerUpdatedWorldChanged(BinaryStream stream) {
        read(stream);
    }

    private PlayerUpdatedWorldChanged(String wn) {
        this.wn = wn;
    }

    @Override
    public boolean equals(Object obj) {
        if (!(obj instanceof PlayerUpdatedWorldChanged o)) return false;
        return (Objects.equals(wn, o.wn));
    }

    @Override
    public void write(BinaryStream stream) {
        stream.putString(wn);
    }

    @Override
    public void read(BinaryStream stream) {
        wn = stream.getString();
    }

    @Override
    public String toString() {
        return "PlayerUpdatedWorldChanged{" +
            "wn='" + wn + '\'' +
            '}';
    }
}

File: src/main/java/net/easecation/ghosty/recording/player/updated/PlayerUpdatedWorldChanged.kt
package net.easecation.ghosty.recording.player.updated

import cn.nukkit.Server
import cn.nukkit.utils.BinaryStream
import kotlinx.serialization.Serializable
import net.easecation.ghosty.entity.PlaybackNPC
import net.easecation.ghosty.recording.UpdateWithState
import net.easecation.ghosty.recording.player.PlayerRecordNode

@Serializable
data class PlayerUpdatedWorldChanged(
    val wn: String
) : PlayerUpdated, UpdateWithState {

    override fun getUpdateTypeId() = PlayerUpdated.TYPE_WORLD

    override fun processTo(ghost: PlaybackNPC) {
        ghost.location.level = Server.getInstance().getLevelByName(wn)
    }

    override fun applyTo(node: PlayerRecordNode) = node.apply { level = wn }

    companion object {
        @JvmStatic
        fun of(wn: String) = PlayerUpdatedWorldChanged(wn)

        @JvmStatic
        val ADAPTER: PlayerUpdateAdapter<PlayerUpdatedWorldChanged> = Adapter
    }

    private object Adapter : PlayerUpdateAdapter<PlayerUpdatedWorldChanged> {
        override fun write(updated: PlayerUpdatedWorldChanged, stream: BinaryStream) {
            stream.putString(updated.wn)
        }

        override fun read(stream: BinaryStream): PlayerUpdatedWorldChanged {
            return PlayerUpdatedWorldChanged(stream.string)
        }
    }
}
